sovrin_node/persistence/idr_cache.py

```diff
@@ -0,0 +1,201 @@
+    the value of role starts. eg.
+    Value in case of guardian: '\2<guardian's DID>\0<role of the DID>'
+    Value in case of no guardian: '\1<verkey of the DID>\0<role of the DID>'
+    """
+
+    unsetVerkey = b'-'
+
+    def __init__(self, basedir: str, name):
+        logger.debug('Initializing identity cache {} at {}'
+                     .format(name, basedir))
+        self._basedir = basedir
+        self._name = name
+        self._db = None
+        self.open()
+        # OrderedDict where key is the state root after batch and value is a
+        # dictionary similar to cache which can be queried like the
+        # database, i.e `self._db`. Keys (state roots are purged) when they
+        # get committed or reverted.
+        self.unCommitted = OrderedDict() # type: Dict[bytes, OrderedDict]
+
+        # Relevant NYMs operation done in current batch, in order
+        self.currentBatchOps = []   # type: List[Tuple]
+
+    @staticmethod
+    def encodeVerkey(verkey):
+        if verkey is None:
+            return IdrCache.unsetVerkey
+        else:
+            if isinstance(verkey, str):
+                return verkey.encode()
+            return verkey
+
+    @staticmethod
+    def decodeVerkey(verkey):
+        if verkey == IdrCache.unsetVerkey:
+            return None
+        else:
+            return verkey.decode()
+
+    @staticmethod
+    def getPrefixAndIv(guardian=None, verkey=None):
+        iv = guardian if guardian else verkey
+        prefix = b'1' if guardian else b'0'
+        return prefix, iv
+
+    @staticmethod
+    def packIdrValue(ta=None, role=None, verkey=None):
+        # prefix, iv = IdrCache.getPrefixAndIv(guardian, verkey)
+        if ta is None:
+            ta = b''
+        if role is None:
+            role = b''
+        verkey = IdrCache.encodeVerkey(verkey)
+        return rlp.encode([ta, verkey, role])
+
+    @staticmethod
+    def unpackIdrValue(value):
+        ta, verkey, role = rlp.decode(value)
+        return ta.decode(), IdrCache.decodeVerkey(verkey), role.decode()
+
+    def get(self, idr, isCommitted=True):
+        idr = idr.encode()
+        if isCommitted:
+            value = self._db.get(idr)
+        else:
+            # Looking for uncommitted values, iterating over `self.unCommitted`
+            # in reverse to get the latest value
+            for i, cache in reversed(self.unCommitted.items()):
+                if idr in cache:
+                    value = cache[idr]
+                    break
+            else:
+                value = self._db.get(idr)
+        ta, iv, r = self.unpackIdrValue(value)
+        return ta, iv, r
+
+    def set(self, idr, ta=None, verkey=None, role=None, isCommitted=True):
+        val = self.packIdrValue(ta, role, verkey)
+        if isCommitted:
+            self._db.set(idr, val)
+        else:
+            self.currentBatchOps.append((idr, val))
+
+    def close(self):
+        self._db.close()
+
+    def open(self):
```

Comment:
If we need to change a KVStore type, we need to change implementation. It's better to inject db, so that using another DB (for example RocksDB) can be done easily without changing IdrCache class.
We can use leveldb as default for db.